osbenchmark/worker_coordinator/runner.py

```diff
@@ -1270,7 +1281,20 @@ def calculate_topk_search_recall(predictions, neighbors, top_k):
                     self.logger.info("No neighbors are provided for recall calculation")
                     return 0.0
                 min_num_of_results = min(top_k, len(neighbors))
+                last_neighbor_is_negative_1 = int(neighbors[min_num_of_results-1]) == -1
```

Comment:
This logic is necessary when the dataset is first generated and then neighbors are filtered out. This occurs in some of the perf tools datasets like [here](https://github.com/opensearch-project/k-NN/blob/main/benchmarks/perf-tool/add-filters-to-dataset.py). These datasets first generate true neighbors for the test vector, such as `[5, 2, 3]`, and then post-filter the neighbors with attributes that do not pass the filter. The documents without the attributes are replaced by `-1`. However looking more closely at the dataset generation code I'm not sure that all of the `-1` entries are at the end of the array. @martin-gaievski I see you wrote the filtering code, could you please clarify? Thanks

Reply:
they should be at the end of collection of nearest neighbors for each query, although I'm not clearly remember sorting them. Can you share example where -1 are in the row, but they are mixed with some other ids

Reply:
I didn't see any in testing, but wanted to ask since I couldn't find a sort. I took another look at the [code for filtering neighbors](https://github.com/opensearch-project/k-NN/blob/a16b8aa965f66268799c001756c78174c7baba19/benchmarks/perf-tool/add-filters-to-dataset.py#L161) and it looks like the initial array with all `-1` entries is overwritten from left to right, so any remaining `-1` ids are at the end. I don't think we'll use these filter datasets anyways (since Heemin's nested fields script generates `k` true neighbors after filtering instead of generating `k` true neighbors and then post-filtering to have `<k` neighbors), but I'll call out the assumption that all `-1` are at the end in a comment.